crates/sui-swarm/src/memory/node.rs

```diff
@@ -0,0 +1,228 @@
+                RuntimeType::SingleThreaded => tokio::runtime::Builder::new_current_thread(),
+                RuntimeType::MultiThreaded => {
+                    thread_local! {
+                        static SPAN: std::cell::RefCell<Option<tracing::span::EnteredSpan>> =
+                            std::cell::RefCell::new(None);
+                    }
+                    let mut builder = tokio::runtime::Builder::new_multi_thread();
+                    let span = span.clone();
+                    builder
+                        .on_thread_start(move || {
+                            SPAN.with(|maybe_entered_span| {
+                                *maybe_entered_span.borrow_mut() = Some(span.clone().entered());
+                            });
+                        })
+                        .on_thread_stop(|| {
+                            SPAN.with(|maybe_entered_span| {
+                                maybe_entered_span.borrow_mut().take();
+                            });
+                        });
+
+                    builder
+                }
+            };
+            let runtime = builder.enable_all().build().unwrap();
+
+            runtime.block_on(async move {
+                let _server = SuiNode::start(&config).await.unwrap();
+                // Notify that we've successfully started the node
+                let _ = startup_sender.send(());
+                // run until canceled
+                cancel_reciever.map(|_| ()).await;
+            });
+        });
+
+        (
+            startup_reciever,
+            Self {
+                thread: Some(thread),
+                cancel_sender: Some(cancel_sender),
+            },
+        )
+    }
+
+    /// Check to see that the Node is still alive by checking if the receiving side of the
+    /// `cancel_sender` has been dropped.
+    ///
+    //TODO When we move to rust 1.61 we should also use
+    // https://doc.rust-lang.org/stable/std/thread/struct.JoinHandle.html#method.is_finished
+    // in order to check if the thread has finished.
+    pub fn is_alive(&self) -> bool {
+        if let Some(cancel_sender) = &self.cancel_sender {
+            !cancel_sender.is_closed()
+        } else {
+            false
+        }
+    }
+}
+
+/// The type of tokio runtime that should be used for a particular Node
+#[derive(Clone, Copy, Debug)]
+pub enum RuntimeType {
+    SingleThreaded,
+    MultiThreaded,
+}
+
+#[cfg(test)]
+mod test {
+    use crate::memory::Swarm;
+
+    #[tokio::test]
+    async fn start_and_stop() {
+        telemetry_subscribers::init_for_testing();
+        let mut swarm = Swarm::builder().build();
+
+        let validator = swarm.validators_mut().next().unwrap();
+
+        validator.start().await.unwrap();
+        validator.health_check().await.unwrap();
+        validator.stop();
+        validator.health_check().await.unwrap_err();
+
+        validator.start().await.unwrap();
+        validator.health_check().await.unwrap();
+    }
```

Comment:
A negative test on a failing node (e.g. with gRPC down) would help, if reasonably not too complicated.

Reply:
I think you missed this, we do have a negative test for when the node itself is down on line 223.

Reply:
If you mean when a node is "up" but with grpc "down" that's not trivial to setup.